src/bin/languageserver/mod.rs

```diff
@@ -650,17 +874,49 @@ impl<'a> Builder<'a> {
                 self.expression(index, symtab);
             }
 
-            ast::Expression::StructMember { loc, ty, expr, .. } => {
+            ast::Expression::StructMember {  loc, expr, field, ty } => {
                 self.expression(expr, symtab);
-                self.hovers.push(HoverEntry {
-                    start: loc.start(),
-                    stop: loc.exclusive_end(),
-                    val: make_code_block(ty.to_string(self.ns)),
-                });
+
+                self.hovers.push((
+                    loc.file_no(),
+                    HoverEntry {
+                        start: loc.start(),
+                        stop: loc.exclusive_end(),
+                        val: make_code_block(ty.to_string(self.ns)),
+                    },
+                ));
+
+                let t = expr.ty().deref_any().clone();
+                if let Type::Struct(StructType::UserDefined(_)) = t {
```

Comment:
I don't think the clone is necessary.
```rust
                if let Type::Struct(StructType::UserDefined(_)) = expr.ty().deref_any() {
```

Reply:
Another alternative would be:
```Rust

if matches!(Type::Struct(StructType::UserDefined(_)), expr.ty().deref_any()) {
```

Reply:
We need `clone` here as we need to pass the type to the `DefinitionType::Field` in line 898. 

Trying to move `clone` to line 898 instead of its current location (line 889), as shown below, leads to another error. 
```rust
let t = expr.ty().deref_any();
if let Type::Struct(StructType::UserDefined(_)) = t {
    self.references.push((
        loc.file_no(),
        ReferenceEntry {
            start: loc.start(),
            stop: loc.exclusive_end(),
            val: DefinitionIndex {
                def_path: Default::default(),
                def_type: DefinitionType::Field(t.clone(), *field),
            },
        },
    ));
}
```

```
error[E0716]: temporary value dropped while borrowed
   --> src/bin/languageserver/mod.rs:889:25
    |
889 |                 let t = expr.ty().deref_any();
    |                         ^^^^^^^^^            - temporary value is freed at the end of this statement
    |                         |
    |                         creates a temporary value which is freed while still in use
890 |                 if let Type::Struct(StructType::UserDefined(_)) = t {
    |                                                                   - borrow later used here
    |
    = note: consider using a `let` binding to create a longer lived value
```
So, in order to prevent this error and to avoid calling the `expr.ty()` twice, I think `clone` is required.

Reply:
Yes, you are right @chioni16. I could not get rid of the clone either (without making the code worse).